Modules/xxlimited.c

```diff
@@ -3,47 +3,103 @@
    also declares object types. All occurrences of 'Xxo' should be changed
    to something reasonable for your objects. After that, all other
    occurrences of 'xx' should be changed to something reasonable for your
-   module. If your module is named foo your sourcefile should be named
-   foomodule.c.
+   module. If your module is named foo your source file should be named
+   foo.c or foomodule.c.
 
    You will probably want to delete all references to 'x_attr' and add
    your own types of attributes instead.  Maybe you want to name your
    local variables other than 'self'.  If your object type is needed in
    other files, you'll have to create a file "foobarobject.h"; see
-   floatobject.h for an example. */
+   floatobject.h for an example.
 
-/* Xxo objects */
+   This module roughly corresponds to::
+
+      class Xxo:
+         """A class that explicitly stores attributes in an internal dict"""
+
+          def __init__(self):
+              # In the C class, "_x_attr" is not accessible from Python code
+              self._x_attr = {}
+
+          def __getattr__(self, name):
+              return self._x_attr[name]
+
+          def __setattr__(self, name, value):
+              self._x_attr[name] = value
+
+          def __delattr__(self, name):
+              del self._x_attr[name]
+
+          def demo(o, /):
+              if isinstance(o, str):
+                  return o
+              elif isinstance(o, Xxo):
+                  return o
+              else:
+                  raise Error('argument must be str or Xxo')
+
+      class Error(Exception):
+          """Exception raised by the xxlimited module"""
+
+      def foo(i: int, j: int, /):
+          """Return the sum of i and j."""
+          # Unlike this pseudocode, the C function will *only* work with
+          # integers and perform C long int arithmetic
+          return i + j
+
+      def new():
+          return Xxo()
+
+      def Str(str):
+          # A trivial subclass of a built-in type
+          pass
+   */
 
 #include "Python.h"
 
-static PyObject *ErrorObject;
+// Module state
+typedef struct {
+    PyObject *Xxo_Type;    // Xxo class
+    PyObject *Error_Type;       // Error class
+} xx_state;
+
+
+/* Xxo objects */
 
+// Instance state
 typedef struct {
     PyObject_HEAD
     PyObject            *x_attr;        /* Attributes dictionary */
 } XxoObject;
 
-static PyObject *Xxo_Type;
-
-#define XxoObject_Check(v)      Py_IS_TYPE(v, Xxo_Type)
+// XXX: no good way to do this yet
+// #define XxoObject_Check(v)      Py_IS_TYPE(v, Xxo_Type)
```

Comment:
It works if you pass a state, and the type from the state, no?

Reply:
I'd like to leave this out of scope for this PR.